plugins/state_history_plugin/include/eosio/state_history_plugin/state_history_log.hpp

```diff
@@ -0,0 +1,275 @@
+      last_block_id = header.block_id;
+   }
+
+   // returns stream positioned at payload
+   std::fstream& get_entry(uint32_t block_num, state_history_log_header& header) {
+      EOS_ASSERT(block_num >= _begin_block && block_num < _end_block, chain::plugin_exception,
+                 "read non-existing block in ${name}.log", ("name", name));
+      log.seekg(get_pos(block_num));
+      log.read((char*)&header, sizeof(header));
+      return log;
+   }
+
+   chain::block_id_type get_block_id(uint32_t block_num) {
+      state_history_log_header header;
+      get_entry(block_num, header);
+      return header.block_id;
+   }
+
+ private:
+   bool get_last_block(uint64_t size) {
+      state_history_log_header header;
+      uint64_t                 suffix;
+      log.seekg(size - sizeof(suffix));
+      log.read((char*)&suffix, sizeof(suffix));
+      if (suffix > size || suffix + sizeof(header) > size) {
+         elog("corrupt ${name}.log (2)", ("name", name));
+         return false;
+      }
+      log.seekg(suffix);
+      log.read((char*)&header, sizeof(header));
+      if (suffix + sizeof(header) + header.payload_size + sizeof(suffix) != size) {
+         elog("corrupt ${name}.log (3)", ("name", name));
+         return false;
+      }
+      _end_block    = header.block_num + 1;
+      last_block_id = header.block_id;
+      if (_begin_block >= _end_block) {
+         elog("corrupt ${name}.log (4)", ("name", name));
+         return false;
+      }
+      return true;
+   }
+
+   void recover_blocks(uint64_t size) {
+      ilog("recover ${name}.log", ("name", name));
+      uint64_t pos       = 0;
+      uint32_t num_found = 0;
+      while (true) {
+         state_history_log_header header;
+         if (pos + sizeof(header) > size)
+            break;
+         log.seekg(pos);
+         log.read((char*)&header, sizeof(header));
+         uint64_t suffix;
+         if (header.payload_size > size || pos + sizeof(header) + header.payload_size + sizeof(suffix) > size)
+            break;
+         log.seekg(pos + sizeof(header) + header.payload_size);
+         log.read((char*)&suffix, sizeof(suffix));
+         if (suffix != pos)
+            break;
+         pos = pos + sizeof(header) + header.payload_size + sizeof(suffix);
+         if (!(++num_found % 10000)) {
+            printf("%10u blocks found, log pos=%12llu\r", (unsigned)num_found, (unsigned long long)pos);
+            fflush(stdout);
+         }
+      }
+      log.flush();
+      boost::filesystem::resize_file(log_filename, pos);
+      log.sync();
+      EOS_ASSERT(get_last_block(pos), chain::plugin_exception, "recover ${name}.log failed", ("name", name));
+   }
+
+   void open_log() {
+      log.open(log_filename, std::ios_base::binary | std::ios_base::in | std::ios_base::out | std::ios_base::app);
+      log.seekg(0, std::ios_base::end);
+      uint64_t size = log.tellg();
+      if (size >= sizeof(state_history_log_header)) {
```

Comment:
After discussion with @tbfleming, it seems there isn't much value for versioning the log structure itself. The intention is not to have to replay this log since that doesn't scale well with a blockchain with a very long history.

It may still be desirable to add version of `state_history_log_header` itself (yes, this would be repeated for each block). This is different than the current `version` field since that refers to the version of the payload.